Mathlib/NumberTheory/FLT/Polynomial.lean

```diff
@@ -0,0 +1,332 @@
+      rcases find_contract hc cd chn0 with ⟨cc, cc_nz, eq_c, eq_deg_c⟩
+      set ch := ringChar k
+      suffices hca : ca.natDegree = 0 by
+        rw [eq_d, eq_deg_a, hca, zero_mul]
+      by_contra hnca; apply hnca
+      apply ih_d _ _ rfl _ ca_nz cb_nz cc_nz <;> clear ih_d <;> try rfl
+      · apply is_coprime_of_expand chn0
+        rw [← eq_a, ← eq_b]
+        exact hab
+      · have _ : ch ≠ 1 := CharP.ringChar_ne_one
+        have hch2 : 2 ≤ ch := by omega
+        rw [← add_le_add_iff_right 1, eq_d, eq_deg_a]
+        refine le_trans ?_ (Nat.mul_le_mul_left _ hch2)
+        omega
+      · rw [eq_a, eq_b, eq_c, ← expand_C ch u, ← expand_C ch v, ← expand_C ch w] at heq
+        simp_rw [← map_pow, ← map_mul, ← map_add] at heq
+        rw [Polynomial.expand_eq_zero (zero_lt_iff.mpr chn0)] at heq
+        exact heq
+
+-- Nonsolvability of Fermat-Catalan equation.
+theorem Polynomial.flt_catalan
+    {p q r : ℕ} (hp : 0 < p) (hq : 0 < q) (hr : 0 < r)
+    (hineq : q * r + r * p + p * q ≤ p * q * r)
+    (chp : ¬ringChar k ∣ p) (chq : ¬ringChar k ∣ q) (chr : ¬ringChar k ∣ r)
+    {a b c : k[X]} (ha : a ≠ 0) (hb : b ≠ 0) (hc : c ≠ 0) (hab : IsCoprime a b)
+    {u v w : k} (hu : u ≠ 0) (hv : v ≠ 0) (hw : w ≠ 0)
+    (heq : C u * a ^ p + C v * b ^ q + C w * c ^ r = 0) :
+    a.natDegree = 0 ∧ b.natDegree = 0 ∧ c.natDegree = 0 := by
+  -- WLOG argument: essentially three times flt_catalan_aux
+  have hbc : IsCoprime b c := by
+    apply rot_coprime heq hab <;> assumption
+  have heq' : C v * b ^ q + C w * c ^ r + C u * a ^ p = 0 := by
+    rw [add_rotate] at heq; exact heq
+  have hca : IsCoprime c a := by
+    apply rot_coprime heq' hbc <;> assumption
+  refine ⟨?_, ?_, ?_⟩
+  · apply flt_catalan_aux heq <;> assumption
+  · rw [add_rotate] at heq hineq
+    rw [mul_rotate] at hineq
+    apply flt_catalan_aux heq <;> assumption
+  · rw [← add_rotate] at heq hineq
+    rw [← mul_rotate] at hineq
+    apply flt_catalan_aux heq <;> assumption
+
+/- FLT is special case of nonsolvability of Fermat-Catalan equation.
+Take p = q = r = n and u = v = 1, w = -1.
+-/
+theorem Polynomial.flt
+    {n : ℕ} (hn : 3 ≤ n) (chn : ¬ringChar k ∣ n)
+    {a b c : k[X]} (ha : a ≠ 0) (hb : b ≠ 0) (hc : c ≠ 0)
+    (hab : IsCoprime a b) (heq : a ^ n + b ^ n = c ^ n) :
+    a.natDegree = 0 ∧ b.natDegree = 0 ∧ c.natDegree = 0 := by
+  have hn' : 0 < n := by linarith
+  rw [← sub_eq_zero, ← one_mul (a ^ n), ← one_mul (b ^ n), ← one_mul (c ^ n), sub_eq_add_neg, ←
+    neg_mul] at heq
+  have hone : (1 : k[X]) = C 1 := by rfl
+  have hneg_one : (-1 : k[X]) = C (-1) := by simp only [map_neg, map_one]
+  simp_rw [hneg_one, hone] at heq
+  apply flt_catalan hn' hn' hn' _
+    chn chn chn ha hb hc hab one_ne_zero one_ne_zero (neg_ne_zero.mpr one_ne_zero) heq
+  have eq_lhs : n * n + n * n + n * n = 3 * n * n := by ring_nf
+  rw [eq_lhs]; rw [mul_assoc, mul_assoc]
+  apply Nat.mul_le_mul_right (n * n); exact hn
```

Comment:
```suggestion
  exact Nat.mul_le_mul_right (n * n) hn
```
I don't like semicolons :-) I think the style is "different tactics on different lines" but here you don't need two tactics.